src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsOperation.cs

```diff
@@ -132,7 +134,114 @@ private IEnumerable<IEdmOperationParameter> ComputeParameters()
                 }
             }
 
-            return parameters;
+            // Handle the out-of-line optinal parameter annotation for parameters.
+            // First, use the above built parameters to create the full target name, for example: NS.TestFunction(Edm.String, Edm.String, Edm.String)
+            // Then, go through each prameters by visiting the out-of-line optional parameter annotation.
+            // If we find at least one of out-of-line optional parameter annotation, we replace it as an optional parameter.
+            // Otherwise, re-use the built parameter.
+            // Be noted: if a parameter has inline and out-of-line optional parameter, the out-of-line will win.
+            string fullName = ParameterizedTargetName(Namespace + "." + Name, parameters);
+
+            List<IEdmOperationParameter> newParameters = new List<IEdmOperationParameter>();
+            foreach (var parameter in parameters)
+            {
+                // for example: NS.TestFunction(Edm.String, Edm.String, Edm.String)/OptionalParameter
+                string fullTargetName = fullName + "/" + parameter.Name;
+                string defaultValue;
+
+                if (TryGetOptionalParameterAnnotation(fullTargetName, out defaultValue))

```

Comment:
Which "IsOptional" property do you mean?  
If you mean the “IsOptional" property on "CsdlOperationParameter", I don't think we should set it (even though we can't set it, except we create a new instance of CsdlOperationParameter.

Reply:
I think what Kanish is suggesting is that, rather than create the (non) optional parameter in line 133, and then loop through the entire collection of parameters calling TryGetOptionalParameterAnnotation on each, we could nest the TryGetOptionalParameterAnnotation in the else on line 131 so that, if the IsOptionalFlag is not set, we look for the external annotation *prior* to creating the CsdlSemanticsOperationParameter, and then just create right one to start with.  i.e., nest lines 149-160 within the else statement on line 131 (except replace newParameters with parameters). 

---
In reply to: [222054166](https://github.com/OData/odata.net/pull/1267#discussion_r222054166) [](ancestors = 222054166)

Reply:
@mikepizzo @KanishManuja-MS Ok. I got it. However, I need the `parameters` created above (foreach) to generate the "fullName" at line 143. See Line 143. Otherwise, it's hard to create the `fullname` because it's hard to get the each parameter type name form the `CsdlOperationParameter`

Reply:
I see.

---
In reply to: [222392493](https://github.com/OData/odata.net/pull/1267#discussion_r222392493) [](ancestors = 222392493)